### ISSUE TITLE
False positive: magicnumbers with bit shift expression

### ISSUE THREAD

**Issue by ghernadi** [checkstyle/checkstyle#15950]
I have read check documentation: https://checkstyle.org/checks/coding/magicnumber.html
I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
$ javac Test.java
$
```

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">
    <module name="MagicNumber">
       <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>
       <property name="ignoreHashCodeMethod" value="true"/>
       <property name="ignoreFieldDeclaration" value="true"/>
       <property name="severity" value="warning"/>
       <property name="id" value="checkstyle:magicnumber"/>
    </module>
  </module>
</module>
```
```java
public class Test {
    public static final int BIT0 = 1;
    public static final int BIT1 = 1 << 1;
    public static final int BIT2 = 1 << 2;
    public static final int BIT3 = 1 << 3;
    public static final int BIT4 = 1 << 4;
}
```
```bash
$ java -jar checkstyle-10.20.1-all.jar -c config.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:5:41: '3' is a magic number. [checkstyle:magicnumber]
[WARN] /var/tmp/Test.java:6:41: '4' is a magic number. [checkstyle:magicnumber]
Audit done.
```

I would have expected bit-shift operators to also be excluded from this check, just like `2 * 2 * 2` would work without checkstyle complaining...

**Comment by nrmancuso** [comment]
Few things:

> I would have expected bit-shift operators to also be excluded from this check,  just like 2 * 2 * 2 would work without checkstyle complaining...

This appears to be a bug; to be clear, the check claims that it:
>Checks that there are no ["magic numbers"](https://en.wikipedia.org/wiki/Magic_number_%28programming%29) where a magic number is a numeric literal that is not defined as a constant. By default, -1, 0, 1, and 2 are not considered to be magic numbers.
Constant definition is any variable/field that has 'final' modifier. It is fine to have one constant defining multiple numeric literals within one expression

So, since these fields are declared as constants and are defined by multiple numeric literal, I would expect that violations are not placed on them

> `<property name="ignoreFieldDeclaration" value="true"/>`

This property being set to true would lead me to believe that NO fields should be checked, regardless; apparently this is not the case. The provided example in the docs confuses matters more: https://checkstyle.org/checks/coding/magicnumber.html#Example3-config

> To configure the check so that it ignores magic numbers in field declarations

> `<property name="ignoreFieldDeclaration" value="false"/>`

>  `private static int myInt = 7; // ok, field declaration`

If ignoreFieldDeclaration is "false", that means that we should NOT be ignoring field declarations :)

Here, I have extended your example to hopefully make the problem a bit more obvious:
```
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="severity" value="error"/>
    <property name="fileExtensions" value="java, properties, xml"/>
    <module name="TreeWalker">
        <module name="MagicNumber">
            <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>
            <property name="ignoreHashCodeMethod" value="true"/>
            <property name="ignoreFieldDeclaration" value="true"/>
            <property name="severity" value="warning"/>
            <property name="id" value="checkstyle:magicnumber"/>
        </module>
    </module>
</module>
                                                                                                  
 ➜  src cat Test.java 
public class Test {
    public static final int BIT0 = 1;
    public static final int BIT1 = 1 << 1;
    public static final int BIT2 = 1 << 2;
    public static final int BIT3 = 1 << 3;
    public static final int BIT4 = 1 << 4;

    public static final int MUL1 = 1;
    public static final int MUL2 = 1 * 1;
    public static final int MUL3 = 1 * 2;
    public static final int MUL4 = 1 * 3;
    public static final int MUL5 = 1 * 4;
}

➜  src javac Test.java 

➜  src java -jar checkstyle-10.17.0-all.jar -c config.xml Test.java
Starting audit...
[WARN] /home/nick/IdeaProjects/tester/src/Test.java:5:41: '3' is a magic number. [checkstyle:magicnumber]
[WARN] /home/nick/IdeaProjects/tester/src/Test.java:6:41: '4' is a magic number. [checkstyle:magicnumber]
Audit done.
```
We should also fix the docs while we are here.

**Comment by mahfouz72** [comment]
I am on it. 

We don't have bitwise operators in `constantWaiverParentToken`. 

> Specify tokens that are allowed in the AST path from the number literal to the enclosing constant definition.

https://github.com/checkstyle/checkstyle/blob/1db5e7eb487d2801475e6d855c691a89fcb0f4f2/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java#L261-L272


when we have an AST with `<<` (`SL`)  in line 266 we don't find it in the `constantWaiverParentToken` so we break and return false.

It is not so obvious for me how we generate this list. I think we are missing some operators in it (bitwise operators for example)

**Comment by mahfouz72** [comment]
This is an expected behaviour according to the check documentation. I don't think this is an issue, The only thing I have here is why we don't expand the default value of `constantWaiverParentToken` to include all tokens, and let the users override it based on their specific needs.

IMO, the default value here is so strict, and makes no sense to me why we chose `STAR` but not `LT` for example

**Comment by romani** [comment]
Look at blame https://github.com/checkstyle/checkstyle/blame/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MagicNumberCheck.java

Updates should reference to issue, and our thoughts on this

**Comment by mahfouz72** [comment]
The commit message: `Made MagicNumber check less aggressive, based on feedback on checkstyle-user list`

looks like it is still aggressive

**Comment by romani** [comment]
Check is very aggressive by design. We are here to find practical cases where it should be less demanding or optionally less demanding.

**Comment by romani** [comment]
Initial set of tokens was 20 years ago.

But few recent:
https://github.com/checkstyle/checkstyle/commit/262bcabc0712b649340a96a7229fba80858eaa03
https://github.com/checkstyle/checkstyle/commit/fabe67975252df0a7e2fe2036e4d50c273745c37

Have tracking of intend. All of them are for fields with expressions.

So I am ok to extend ignore list with binary operations. We will see in regression diff report , how it affects others.

Please send PR.

**Comment by romani** [comment]
@ghernadi,
workaround is to add `SR, BSR, SL, BXOR, BOR, BAND, BNOT` to `constantWaiverParentToken` property. Or any other AST tokens you need to ignore.

```
$ cat Test.java 
public class Test {
    public static final int BIT0 = 1;
    public static final int BIT1 = 1 << 1;
    public static final int BIT2 = 1 << 2;
    public static final int BIT3 = 1 << 3;
    public static final int BIT4 = 1 << 4;
}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">
    <module name="MagicNumber">
       <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>
       <property name="ignoreHashCodeMethod" value="true"/>
       <property name="ignoreFieldDeclaration" value="true"/>
       <property name="severity" value="warning"/>
       <property name="id" value="checkstyle:magicnumber"/>
       <property name="constantWaiverParentToken" value="ARRAY_INIT , ASSIGN , DIV ,
        ELIST , EXPR , LITERAL_NEW , METHOD_CALL , MINUS ,
        PLUS , STAR , TYPECAST , UNARY_MINUS , UNARY_PLUS,
        SR,
        BSR,
        SL,
        BXOR,
        BOR,
        BAND,
        BNOT"/>

    </module>
  </module>
</module>

$ java -jar checkstyle-10.20.2-all.jar -c config.xml Test.java 
Starting audit...
Audit done.
```

in referenced PR we just consider to add such tokens by default or not.

**Comment by romani** [comment]
@mahfouz72 , do you think that in scope of this issue we should just make `ignoreFieldDeclaration=true` to ignore all possible numbers in expression part of filed declaration?

Let's put on a side question about default set of tokens for constantWaiverParentToken.

**Comment by mahfouz72** [comment]
>  do you think that in scope of this issue we should just make ignoreFieldDeclaration=true to ignore all possible numbers in expression part of filed declaration?

Yes, regardless of the question about the default value of `constantWaiverParentToken` the provided code in the issue description should not be a violation.

> ignoreFieldDeclaration: Ignore magic numbers in field declarations.   (Docs)

IMO, when `ignoreFieldDeclaration` is set to true, we should give this property the highest priority and strictly ignore all fields regardless of the operators and numbers appearing in them.

---

> What I do not understand is why this token https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ASSIGN does NOT always match, as constant is always has =.

It is always a match that is why we have `ASSIGN` in `constantWaiverParentToken` 

> Specify tokens that are allowed in the AST path from the number literal to the enclosing constant definition.

`=` is allowed because it is always a part of a constant def

---

Ok, to keep things organized we should
1. Close the current linked PR that tries to expand the default list
2. Move this decision (expanding default value of `constantWaiverParentToken`) in another issue and vote on what is best to do
3.  The scope of this issue will be to fix the violation in the OP (That is when `ignoreFieldDeclaration=true` we should always ignore fields regardless of the numbers, operators,  and constant field or not)

**Comment by nrmancuso** [comment]
> Ok, to keep things organized we should

Close the current linked PR that tries to expand the default list
Move this decision (expanding default value of constantWaiverParentToken) in another issue and vote on what is best to do
The scope of this issue will be to fix the violation in the OP (That is when ignoreFieldDeclaration=true we should always ignore fields regardless of the numbers, operators, and constant field or not)

This makes sense to me, @romani you good?

**Comment by romani** [comment]
>The scope of this issue will be to fix the violation in the OP (That is when ignoreFieldDeclaration=true we should always ignore fields regardless of the numbers, operators, and constant field or not)

Yes. Issue is approved.
@mahfouz72, do you have bandwidth to fix this issue?

**Comment by mahfouz72** [comment]
Unfortunately, I am packed this week, I will pick it up as soon as I can

**Comment by romani** [comment]
@mahfouz72, did you fix this issue? I am lost tracking why it is still open.

**Comment by mahfouz72** [comment]
> did you fix this issue? I am lost tracking why it is still open.

@romani No, not yet

The violation is gone as a side effect of expanding the default list with bitwise operators in my last PR. (https://github.com/checkstyle/checkstyle/pull/15957)

But we still have a bug here that we agreed before that this should be no violation even without extending the default list because `<property name="ignoreFieldDeclaration" value="true"/>`  should have a higher priority


> The scope of this issue will be to fix the violation in the OP (That is when ignoreFieldDeclaration=true we should always ignore fields regardless of the numbers, operators, and constant field or not)

---

could you please update the OP with this example to make it super clear what is the bug here?


```
D:\CS\test cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">
    <module name="MagicNumber">
       <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>
       <property name="ignoreHashCodeMethod" value="true"/>
       <property name="ignoreFieldDeclaration" value="true"/>
        <property name="constantWaiverParentToken" value="ARRAY_INIT , ASSIGN  , ELIST , EXPR"/>
    </module>
  </module>
</module>
D:\CS\test cat src/Test.java
public class Test {
    public final int radius = 10;
    public final double area = (22 / 7.0) * radius * radius;
    // 2 violations above, but it should be ok because
    // ignoreFieldDeclaration = true
}
D:\CS\test java  -jar checkstyle-10.21.1-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:3:33: '22' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:3:38: '7.0' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 2 errors.
```

Although constantWaiverParentToken doesn't have `STAR` or `DIV` in the config. but still there should be no violation because these magic numbers are in a field declaration and `ignoreFieldDeclaration` is true

**Comment by romani** [comment]
Please move your comment to new issue. To state actual problem very clearly without any history.

**Comment by mahfouz72** [comment]
https://github.com/checkstyle/checkstyle/issues/16101
